File: September2022/Java/TrappingRainWater.java

// @saorav21994
// TC : O(n)
// SC : O(1)

class Solution {
    public int trap(int[] height) {
        int l = 0, r = height.length - 1;
        int maxl = 0, maxr = 0, res = 0;
        while (l < r) {
            if (height[l] < height[r]) {
                maxl = Math.max(maxl, height[l]);
                res += (maxl-height[l]);
                l += 1;
            }
            else {
                maxr = Math.max(maxr, height[r]);
                res += (maxr-height[r]);
                r -= 1;
            }
        }
        return res;
    }
}

File: September2022/C++/TrappingRainWater.cpp
// https://leetcode.com/problems/trapping-rain-water/
class Solution {
public:
	// TC: O(n)
	// SC: O(1)
    int trap(vector<int>& height) {
        int left = 0, right = height.size() - 1;
        int maxl = 0, maxr = 0, res = 0;
        
        while (left < right) {
            if (height[left] < height[right]) {
                maxl = max(maxl, height[left]);
                res += (maxl - height[left]);
                left += 1;
            } else {
                maxr = max(maxr, height[right]);
                res += (maxr - height[right]);
                right -= 1;
            }
        }
        
        return res;
    }
};
